### PR TITLE
update sdk for rust

### PR DESCRIPTION
**Pull request by ForLisiteng**
1. Modified CreateKey, Sign and other interface parameter.
2. Fix is_base function.
3. Modified unit test.

### CODE REVIEW

File: sdk/rust/src/api.rs [hunk 1]
```diff
@@ -52,28 +41,75 @@ keyid    -- A uinque keyid of the cmk.
...
+        -EH_DIGEST(users need to fill in a digest value calculated using the digest mode.)
+
+padding_mode    --padding_mode is necessary when keyspec is RSA.
+        -EH_RSA_PKCS1(only support rsa keyspec)
+        -EH_RSA_PKCS1_PSS(only support rsa keyspec)
+        -EH_PAD_NONE(supprot ecc and sm2 keyspec)
+
+digest_mode     --If digest mode is not provided, the default digest mode will be used.
+                  EH_SHA_SHA256 will be used for rsa and ecc, and EH_SM3 will be used for sm2.
+				  If use sm2 keypair, digest mode must be EH_SM3.
```
Comment (syan10) [inline sdk/rust/src/api.rs:62]: the indent is inconsistent

File: sdk/rust/src/api.rs [hunk 2]
```diff
@@ -52,28 +41,75 @@ keyid    -- A uinque keyid of the cmk.
...
```
Comment (syan10) [inline sdk/rust/src/api.rs:96]: same here

File: sdk/rust/src/kms.rs [hunk 3]
```diff
@@ -111,11 +102,20 @@ impl KMS for EHSMClient {
...
+        message_type: &str, 
+        message: &str) -> Result<String> {
         let mut payload: Map<String, Value> = Map::new();
 
         payload.insert("keyid".to_owned(), Value::String(keyid.to_owned()));
-        payload.insert("digest".to_owned(), Value::String(digest.to_owned()));
+        payload.insert("padding_mode".to_owned(), Value::String(padding_mode.to_owned()));
+        payload.insert("digest_mode".to_owned(), Value::String(digest_mode.to_owned()));
+        payload.insert("message_type".to_owned(), Value::String(message_type.to_owned()));
+        payload.insert("message".to_owned(), Value::String(message.to_owned()));
```
Comment (syan10) [inline sdk/rust/src/kms.rs:118]: base64 check?

File: sdk/rust/src/kms.rs [hunk 4]
```diff
@@ -130,11 +130,21 @@ impl KMS for EHSMClient {
...
+        message: &str, 
+        signature: &str) -> Result<bool> {
         let mut payload: Map<String, Value> = Map::new();
 
         payload.insert("keyid".to_owned(), Value::String(keyid.to_owned()));
-        payload.insert("digest".to_owned(), Value::String(digest.to_owned()));
+        payload.insert("padding_mode".to_owned(), Value::String(padding_mode.to_owned()));
+        payload.insert("digest_mode".to_owned(), Value::String(digest_mode.to_owned()));
+        payload.insert("message_type".to_owned(), Value::String(message_type.to_owned()));
+        payload.insert("message".to_owned(), Value::String(message.to_owned()));
```
Comment (syan10) [inline sdk/rust/src/kms.rs:147]: same

File: sdk/rust/src/test.rs [hunk 5]
```diff
@@ -8,52 +8,42 @@ mod tests {
...
 
     #[tokio::test]
-    async fn test_asymmetrickey_generate_key() {
+    async fn test_generate_key() {
         let mut client = EHSMClient::new();
 
-        let keyspecs = vec!["EH_RSA_3072", "EH_RSA_4096", "EH_RSA_2048", "EH_SM2", "EH_EC_P224", "EH_EC_P256", "EH_EC_P384", "EH_EC_P521"];
-        let padding_modes = vec![Some("EH_PAD_RSA_PKCS1_OAEP"), Some("EH_PAD_RSA_PKCS1"), None];
-        let digest_modes = vec![Some("EH_SHA_2_224"), Some("EH_SHA_2_256"), Some("EH_SHA_2_384"), Some("EH_SHA_2_512")];
+        let keyspecs = vec!["EH_RSA_2048", "EH_RSA_3072", "EH_RSA_4096", "EH_SM2", "EH_EC_P224", "EH_EC_P256", "EH_EC_P384", "EH_EC_P521", "EH_AES_GCM_128", "EH_AES_GCM_192", "EH_AES_GCM_256", "EH_SM4_CTR", "EH_SM4_CBC"];
```
Comment (syan10) [inline sdk/rust/src/test.rs:14]: code is too long, please change the line

File: sdk/rust/src/test.rs [hunk 6]
```diff
@@ -8,52 +8,42 @@ mod tests {
...
+        let keyusage = vec!["EH_KEYUSAGE_ENCRYPT_DECRYPT", "EH_KEYUSAGE_SIGN_VERIFY"]
         for keyspec in keyspecs.iter() {
```
Comment (syan10) [inline sdk/rust/src/test.rs:16]: EH_KEYUSAGE_SIGN_VERIFY can't use for symmetric key, and does we have check in the low level implementation?

Comment (syan10) [reply sdk/rust/src/test.rs:16]: if not, then we need to add it @whh39

File: sdk/rust/src/test.rs [hunk 7]
```diff
@@ -8,52 +8,42 @@ mod tests {
...
     async fn test_get_publickey() {
         let mut client = EHSMClient::new();
-
-        let keyid = client.create_key("EH_SM2", "EH_INTERNAL_KEY", None, None, None)
-                            .await  
-                            .expect("fail to get keyid");
-        let pubkey = client.get_publickey(&keyid.to_owned()).await;
-
-        assert!(pubkey.is_ok());
+        let keyspecs = vec!["EH_RSA_2048", "EH_RSA_3072", "EH_RSA_4096", "EH_SM2", "EH_EC_P224", "EH_EC_P256", "EH_EC_P384", "EH_EC_P521", "EH_AES_GCM_128", "EH_AES_GCM_192", "EH_AES_GCM_256", "EH_SM4_CTR", "EH_SM4_CBC"];
```
Comment (syan10) [inline sdk/rust/src/test.rs:36]: same here, change the line

Comment (syan10) [reply sdk/rust/src/test.rs:36]: also for the symmetric key, it does not support to getpubkey.

File: sdk/rust/src/test.rs [hunk 8]
```diff
@@ -63,7 +53,7 @@ mod tests {
         let decodedata = &encode(PLAINT_TEXT).to_string()[..];
         let keyspecs = vec!["EH_AES_GCM_128", "EH_AES_GCM_192", "EH_AES_GCM_256", "EH_SM4_CTR"];
         for keyspec in keyspecs.iter() {
-            let keyid = client.create_key(keyspec, "EH_INTERNAL_KEY",None, None, None)
+            let keyid = client.create_key(keyspec, "EH_INTERNAL_KEY", "EH_KEYUSAGE_ENCRYPT_DECRYPT")
                         .await
```
Comment (syan10) [inline sdk/rust/src/test.rs:57]: please add a negative case:  for symmetric key input a EH_KEYUSAGE_SIGN_VERIFY which expect failed

File: sdk/rust/src/client.rs [hunk 9]
```diff
@@ -95,25 +96,29 @@ pub async fn do_get(base_url: &str, action: &str) -> Result<Value, Box<dyn std::
...
  
-        && base64::decode_config(s, base64::STANDARD).is_ok()
+//         && base64::decode_config(s, base64::STANDARD).is_ok()
  
- }
+// }
+pub fn is_base64(s: &str) -> bool {
+	decode(s).unwrap();
+	return true
+}
```
Comment (syan10) [inline sdk/rust/src/client.rs:121]: fn is_base64(s: &str) -> bool {
    base64::decode_config(s, base64::STANDARD).is_ok()
}

File: sdk/rust/src/kms.rs [hunk 10]
```diff
@@ -130,12 +178,46 @@ impl KMS for EHSMClient {
...
+            payload.insert("digest_mode".to_owned(), Value::String(digest_mode.to_owned()));
+        } else {
+            return Err::<bool, anyhow::Error>(anyhow::Error::msg("digest_mode is empty."));
+        }
+        if message_type != "" {
+            payload.insert("message_type".to_owned(), Value::String(message_type.to_owned()));
+        } else {
+            return Err::<bool, anyhow::Error>(anyhow::Error::msg("message_type is empty."));
+        }
+        if crate::client::is_base64(message) {
```
Comment (syan10) [inline sdk/rust/src/kms.rs:211]: msg != ""

File: sdk/rust/src/kms.rs [hunk 11]
```diff
@@ -67,16 +76,24 @@ impl KMS for EHSMClient {
...
                 payload.insert("aad".to_owned(), Value::String(aad_b64.to_owned()));
             }
+        } else {
+            return Err::<String, anyhow::Error>(anyhow::Error::msg("add is false."));
+        }
+        if crate::client::is_base64(data_b64) && data_b64 != "" {
             payload.insert("plaintext".to_owned(), Value::String(data_b64.to_owned()));
         } else {
-            println!("Please base64 encode the value");
+            return Err::<String, anyhow::Error>(anyhow::Error::msg("data is false."));
```
Comment (syan10) [inline sdk/rust/src/kms.rs:96]: refine the error log

File: sdk/rust/src/kms.rs [hunk 12]
```diff
@@ -67,16 +76,24 @@ impl KMS for EHSMClient {
...
+            payload.insert("keyid".to_owned(), Value::String(keyid.to_owned()));
+        } else {
+            return Err::<String, anyhow::Error>(anyhow::Error::msg("Keyid is empty."));
+        }
+        if crate::client::is_base64(aad_b64.unwrap()) && aad_b64 != Some("") {
             if let Some(aad_b64) = aad_b64 {
                 payload.insert("aad".to_owned(), Value::String(aad_b64.to_owned()));
             }
+        } else {
+            return Err::<String, anyhow::Error>(anyhow::Error::msg("Add should be base64 encrypted and not empty."));
```
Comment (syan10) [inline sdk/rust/src/kms.rs:91]: aad can be empty

File: sdk/rust/Cargo.toml [hunk 13]
```diff
@@ -18,6 +18,7 @@ url = "2.2"
 hmac = "0.9"
 sha2 = "0.9"
 rand = "0.3.17"
+rust-crypto = "0.2.36"
```
Comment (syan10) [inline sdk/rust/Cargo.toml:22]: version is too old, may need to be updated

File: sdk/rust/src/test.rs (hunk 6 above)
Comment (whh39) [reply sdk/rust/src/test.rs:14]: I add it on https://github.com/intel/ehsm/pull/298

**Review by syan10 — APPROVED** [review]

**Pull request merged**